Run testbed on fixed Ubuntu version and corresponding system Python version

mhsmith opened the pull request:
Similar to https://github.com/beeware/briefcase/pull/1130.

## PR Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
- [x] All new features have been tested
- [x] All new features have been documented
- [x] I have read the **CONTRIBUTING.md** file
- [x] I will abide by the code of conduct

mhsmith commented:
@rmartin16 @freakboy3742: Any idea what's going on with this CI failure? I can't reproduce it within Docker on my Linux machine, either on debian:bullseye or ubuntu:jammy.
```
[testbed] Starting test_suite...
===========================================================================
Install path: /home/runner/work/toga/toga/testbed/build/testbed/ubuntu/jammy/testbed-0.0.1/usr
Pre-initializing Python runtime...
PYTHONPATH:
- /usr/lib/python3.10
- /usr/lib/python3.10/lib-dynload
- /home/runner/work/toga/toga/testbed/build/testbed/ubuntu/jammy/testbed-0.0.1/usr/lib/testbed/app
- /home/runner/work/toga/toga/testbed/build/testbed/ubuntu/jammy/testbed-0.0.1/usr/lib/testbed/app_packages
Configure argc/argv...
Initializing Python runtime...
Running app module: tests.testbed
---------------------------------------------------------------------------
Application quit abnormally (Exit code -6)!
Traceback (most recent call last):
  File "/usr/lib/python3.10/runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/lib/python3.10/runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "/home/runner/work/toga/toga/testbed/build/testbed/ubuntu/jammy/testbed-0.0.1/usr/lib/testbed/app/tests/testbed.py", line 3, in <module>
    import tempfile
  File "/usr/lib/python3.10/tempfile.py", line 184, in <module>
    from random import Random as _Random
  File "/usr/lib/python3.10/random.py", line 49, in <module>
    from math import log as _log, exp as _exp, pi as _pi, e as _e, ceil as _ceil
ModuleNotFoundError: No module named 'math'
```

freakboy3742 commented:
@mhsmith Mechanically, it looks like it's not finding the Python standard library - except that it is finding *some* of the standard library (`tempfile`). Since it's the math module that is missing, my guess is that it's not finding a binary module; those *should* be in /usr/bin/python3.10/lib-dynload, AFAIK, but either it's not present on the CI machine (possibly due to no having installed the necessary stdlib package?), or the path is different. Unfortunately, my Linux test machine is back at home, so I can't test that hypothesis easily.

rmartin16 commented:
Ohh...user error 🤦🏼‍♂️. I was calling `briefcase build linux` separately from `briefcase run linux --test`....so `build` didn't install the `test_requires` packages and `run` just ran what was already built.

Anyway, using `unset LD_LIBRARY_PATH` before the `briefcase run linux --test` was enough to get things working.

rmartin16 commented:
Also, since you're in CI...I discovered this little snippet to help make CI a little more efficient:
```
concurrency:
  group: ${{ github.ref }}
  cancel-in-progress: true
```
If you add that to the top level, it'll automatically cancel any current CI runs for a PR for a new run is triggered. Useful for quick successive pushes to a branch.

freakboy3742 commented:
> Trying to understand now why the `test_requires` packages seemingly aren't being installed....

If you ran the app previously in non-test mode, it won't notice that dependencies have been "updated". A `-r` pass will fix that.

mhsmith commented:
> Anyway, using `unset LD_LIBRARY_PATH` before the `briefcase run linux --test` was enough to get things working.

Thanks very much. I don't know why LD_LIBRARY_PATH would be causing a problem with the `setup-python`-installed Python, but on further thought, we should be using the Ubuntu system Python package anyway, so I've changed the CI to do that instead.

mhsmith commented:
I've also changed the system package list slightly, and made it identical in both the CI and the tutorial.

I had to add python3-cairo-dev to deal with [this error](https://github.com/beeware/toga/actions/runs/4413499302/jobs/7734065908):
```
      Package py3cairo was not found in the pkg-config search path.
      Perhaps you should add the directory containing `py3cairo.pc'
      to the PKG_CONFIG_PATH environment variable
      No package 'py3cairo' found
      Command '('pkg-config', '--print-errors', '--exists', 'py3cairo >= 1.16.0')' returned non-zero exit status 1.
```

And I removed the following redundant requirements:
* python3-gi is required by python3-gi-cairo
* gir1.2-gtk-3.0 and libwebkit2gtk-4.0-37 are required by gir1.2-webkit2-4.0

freakboy3742 commented on .github/workflows/ci.yml:
```diff
@@ -189,14 +189,20 @@ jobs:
-            runs-on: ubuntu-latest
-            pre-command: "sudo apt-get update -y && sudo apt-get install -y python3-gi python3-gi-cairo gir1.2-gtk-3.0 python3-dev libgirepository1.0-dev libcairo2-dev pkg-config libfuse2"
+            runs-on: ubuntu-22.04
+            # The package list should be the same as in tutorial-0.rst.
```
This list also exists in Toga README, and the BeeWare tutorial (docs.beeware.org) step 0. AFAICT, all three should be in sync.

mhsmith replied:
OK, I've made the README reference the Toga tutorial, and created https://github.com/beeware/beeware/pull/196 for the BeeWare tutorial.

freakboy3742 commented on .github/workflows/ci.yml:
```diff
@@ -189,14 +189,20 @@ jobs:
-            pre-command: "sudo apt-get update -y && sudo apt-get install -y python3-gi python3-gi-cairo gir1.2-gtk-3.0 python3-dev libgirepository1.0-dev libcairo2-dev pkg-config libfuse2"
+            runs-on: ubuntu-22.04
+            # The package list should be the same as in tutorial-0.rst.
+            pre-command: "sudo apt-get update -y && sudo apt-get install -y python3-dev python3-cairo-dev python3-gi-cairo libgirepository1.0-dev libcairo2-dev libpango1.0-dev gir1.2-webkit2-4.0 pkg-config libfuse2"
```
I'm fairly certain libfuse2 was only needed because we were building AppImages - It should be possible to drop it now that we're testing system packages.

mhsmith replied:
OK, done.

freakboy3742 commented on docs/tutorial/tutorial-0.rst:
```diff
@@ -74,7 +77,7 @@ Next, install Toga into your virtual environment:
       # Ubuntu 18.04+ / Debian 10+
       (venv) $ sudo apt-get update
-      (venv) $ sudo apt-get install python3-dev python3-gi python3-gi-cairo libgirepository1.0-dev libcairo2-dev libpango1.0-dev libwebkit2gtk-4.0-37 gir1.2-webkit2-4.0
+      (venv) $ sudo apt-get install python3-dev python3-cairo-dev python3-gi-cairo libgirepository1.0-dev libcairo2-dev libpango1.0-dev gir1.2-webkit2-4.0 pkg-config libfuse2
```
As above, syncing with other sources, without libfuse2.

freakboy3742 reviewed (changes requested):
A couple of notes inline, but mostly makes sense. 

The only change that strikes me as weird is the python-cairo-dev dependency - CI is installing pycairo from a wheel, which I *thought* would mean that we wouldn't need a system package for Cairo, just the runtime one.

mhsmith commented:
> The only change that strikes me as weird is the python-cairo-dev dependency - CI is installing pycairo from a wheel, which I _thought_ would mean that we wouldn't need a system package for Cairo, just the runtime one.

The failure ([full log](https://github.com/beeware/toga/actions/runs/4413499302/jobs/7734065908)) was actually when building pygobject. I'm not sure why it failed here when it was working on the [last merge to the main branch](https://github.com/beeware/toga/actions/runs/4335686681/jobs/7570492777)): maybe because we moved from Python 3.11 to 3.10, or from `setup-python` to the system Python package. But in this case, I actually could reproduce the error locally, so it looks like it's a genuine requirement.

rmartin16 commented:
> > The only change that strikes me as weird is the python-cairo-dev dependency - CI is installing pycairo from a wheel, which I _thought_ would mean that we wouldn't need a system package for Cairo, just the runtime one.
> 
> The failure ([full log](https://github.com/beeware/toga/actions/runs/4413499302/jobs/7734065908)) was actually when building pygobject. I'm not sure why it failed here when it was working on the [last merge to the main branch](https://github.com/beeware/toga/actions/runs/4335686681/jobs/7570492777)): maybe because we moved from Python 3.11 to 3.10, or from `setup-python` to the system Python package. But in this case, I actually could reproduce the error locally, so it looks like it's a genuine requirement.

FWIW, I tracked this down. The `pygobject` package pokes around to [find](https://gitlab.gnome.org/GNOME/pygobject.git) `pycairo` during installation. Since `pycairo` is listed as a build requirement (as well as a runtime requirement), `pip` puts it the isolated build environment before trying to build `pygobject`. Normally, this is where `pygobject` will find `pycairo`.

However, when you're using the System Python, then `/usr/lib/python3/dist-packages` comes in to play. Since we're installing `python3-gi-cairo`, the use of `importlib.utils.find_spec()` returns `/usr/lib/python3/dist-packages/cairo`. However, this is a minimal install and doesn't include the header file. When `python3-gi-cairo` isn't installed, the `pycairo` that was installed in to the `pip` build environment for `pygobject` is found and everything successfully builds.

Furthermore, I was able to run the testbed CI as well as run the testbed app locally without installing `python3-gi-cairo`.

freakboy3742 reviewed (approved):
I've made a small tweak to the README markup adding a link, rather than a raw path; but otherwise 👍

rmartin16 commented:
> > Anyway, using `unset LD_LIBRARY_PATH` before the `briefcase run linux --test` was enough to get things working.
> 
> Thanks very much. I don't know why LD_LIBRARY_PATH would be causing a problem with the `setup-python`-installed Python, but on further thought, we should be using the Ubuntu system Python package anyway, so I've changed the CI to do that instead.

Closing this loop....when `actions/setup-python` installs Python, it [sets](https://github.com/actions/setup-python/blob/dfa76f8d0df0f92390ab2d42e57c6537748ecc71/src/find-python.ts#LL111C11-L111C11) several env vars including `LD_LIBRARY_PATH`. So, when the stub runs, it pulls the Python library from `setup-python` instead of the `/usr/lib` libraries it was built against. I'm not sure of all the expectations in this situation but avoiding it did resolve the issue for me.

The `actions/setup-python` action supports the `update-environment` boolean input to disable taking over an environment like this.....although, avoiding it altogether makes more sense here.
